larq_compute_engine/mlir/transforms/optimize.cc

```diff
@@ -58,12 +68,84 @@ DenseElementsAttr Bitpack(PatternRewriter& builder, Attribute x) {
 
 #include "larq_compute_engine/mlir/transforms/generated_optimize.inc"
 
+struct SetBconvReadWriteBitpacked : public OpRewritePattern<TF::LceBconv2dOp> {
+  using OpRewritePattern<TF::LceBconv2dOp>::OpRewritePattern;
+
+  PatternMatchResult matchAndRewrite(TF::LceBconv2dOp bconv_op,
+                                     PatternRewriter& rewriter) const override {
+    Value bconv_input = bconv_op.input();
+    if (!bconv_input.hasOneUse()) return matchFailure();
+
+    auto bconv_input_op =
+        dyn_cast_or_null<TF::LceBconv2dOp>(bconv_input.getDefiningOp());
+    if (!bconv_input_op) return matchFailure();
+
+    if (bconv_input_op.write_bitpacked_output() &&
+        bconv_op.read_bitpacked_input())
+      return matchFailure();
+
+    if (bconv_input_op.padding() == "SAME" && bconv_input_op.pad_values() == 0)
+      return matchFailure();
+
+    const auto inner_tensor_shape =
+        bconv_input_op.getType().cast<ShapedType>().getShape();
+    if (inner_tensor_shape.size() != 4) return matchFailure();
+
+    // We use 32-bit bitpacking.
+    constexpr int bitwidth = 32;
+
+    const auto channels = inner_tensor_shape[3];
+    const auto packed_channels = (channels + bitwidth - 1) / bitwidth;
+
+    RankedTensorType inner_tensor_type =
+        RankedTensorType::get({inner_tensor_shape[0], inner_tensor_shape[1],
+                               inner_tensor_shape[2], packed_channels},
+                              rewriter.getIntegerType(bitwidth));
+
+    rewriter.replaceOpWithNewOp<TF::LceBconv2dOp>(
+        bconv_input_op, inner_tensor_type, bconv_input_op.input(),
+        bconv_input_op.filter(), bconv_input_op.post_activation_multiplier(),
+        bconv_input_op.post_activation_bias(),
+        rewriter.getIntegerAttr(rewriter.getIntegerType(32),
+                                bconv_input_op.channels_in()),
+        bconv_input_op.strides(),
+        rewriter.getStringAttr(bconv_input_op.padding()),
```

Comment:
Doesn't `bconv_input_op.padding()` work? Why do we need to use `rewriter.getStringAttr`here?

Reply:
I'm not sure specifically why it's necessary, but it doesn't compile without it.

Reply:
Hmm OK, let's leave it like this then :+1: